docs/adr/adr-11-blocksync-overhaul.md

```diff
@@ -0,0 +1,329 @@
+Merkle Proofs serializes into CARv1 file. All the serialized CARv1 file blobs are stored as OS FS files as DAGStore
+[Mounts](https://github.com/filecoin-project/dagstore/blob/master/mount/mount.go).
+
+The introduced `EDSStore` also maintains (via DAGStore) a top-level index enabling granular and efficient random access
+to every share and/or Merkle proof over every registered CARv1 file. The `EDSStore` provides a custom `Blockstore` interface
+implementation to achieve the access. The main use-case is randomized sampling over the whole chain of EDS block data
+and getting data by namespace.
+
+EDSStore constructor should instantiate
+
+```go
+type EDSStore struct {
+ basepath string 
+ dgstr dagstore.DAGStore
+ topIdx index.Inverted 
+ carIdx index.FullIndexRepo
+ mounts  *mount.Registry
+ ...
+}
+
+// NewEDSStore constructs EDStore over OS directory to store CARv1 files of EDSes and indices for them.
+// Datastore is used to keep inverted/top-level index.
+func NewEDSStore(basepath string, ds datastore.Batching) *EDSStore {
+ topIdx := index.NewInverted(datastore)
+ carIdx := index.NewFSRepo(basepath + "/index")
+ mounts := mount.NewRegistry()
+ return &EDSStore{
+  basepath: basepath,
+  dgst: dagstore.New(dagstore.Config{...}),
+  topIdx: index.NewInverted(datastore),
+  carIdx: index.NewFSRepo(basepath + "/index")
+  mounts: mounts,
+    }   
+}
+
+```
+
+##### `share.EDSStore.Put`
+
+To write an entire EDS `Put` method is introduced. Internally it
+
+- Opens a file under `storepath/DataRoot` path
+- Serializes the EDS into the file via `share.WriteEDS`
+- Wraps it with `DAGStore`'s [FileMount](https://github.com/filecoin-project/dagstore/blob/master/mount/file.go#L10)
+- Converts `DataRoot` into the [`shard.Key`](https://github.com/filecoin-project/dagstore/blob/master/shard/key.go#L12)
+- Registers the `Mount` as a `Shard` on the `DAGStore`
+  - This register the `Mount` in [`mount.Registry`](https://github.com/filecoin-project/dagstore/blob/master/mount/registry.go#L22)
+passed to `DAGStore`.
+
+NOTE: Registering on the DAGStore automatically populates top-level index with shares/proofs accessible from stored EDS
+so this is out of scope of the document.
+
+```go
+// Put stores the given data square with DataRoot as key.
+//
+// The square is verified on the Exchange level and Put only stores the square trusting it.
+// The resulting file stores all the shares and NMT Merkle Proofs of the EDS.
+// Additionally, the file gets indexed s.t. Store.Blockstore can access them. 
+func (s *Store) Put(context.Context, DataRoot, *rsmt2d.ExtendedDataSquare) error
+```
+
+##### `share.EDSStore.GetCAR`
+
+To read an EDS as a byte stream `GetCAR` method is introduced. Internally it
+
+- Converts `DataRoot` into the [`shard.Key`](<https://github.com/filecoin-project/dagstore/blob/master/shard/key.go#L12>
+- Gets Mount by `shard.Key` from [`mount.Registry`](https://github.com/filecoin-project/dagstore/blob/master/mount/registry.go#L22)
```

Comment:
Hmm, where does AcquireShard returns a [Reader](https://github.com/filecoin-project/dagstore/blob/master/dagstore.go#L435). There is a result channel, but it only returns ShardAccessor which is not the Reader but a Store basically.

Reply:
Thinking if you really need the reader here we should just extend the interface to return it

Reply:
> Thinking if you really need the reader here we should just extend the interface to return it

The whole purpose of `GetCAR` is to get a Reader 😁. Extending the interface is the way to go if it's allowed to acquire shard for multiple readers.
---
Mounts are just DAGStore native abstraction to get such a Reader. I've decided to stick to those instead of keeping os.File.

Reply:
Updating this section

Reply:
ping me when updated

Reply:
Ping

Reply:
Looks great now